drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,335 @@
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	pad_conf = pad_configs[(8 * pin_cfg) + gpio_num];
+	gpio_cfg.xPadConf = &pad_conf;
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Reset IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~data->gpio_irq_num[gpio_num];
+	} else {
+		/* Prepare configuration */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			gpio_cfg.intr_type = LEVEL_TRIGGERED;
+			if (trig == GPIO_INT_TRIG_LOW) {
+				gpio_cfg.pol_type = FALL_LOW;
+			} else {
+				gpio_cfg.pol_type = RISE_HIGH;
+			}
+		} else {
+			gpio_cfg.intr_type = EDGE_TRIGGERED;
+			switch (trig) {
+			case GPIO_INT_TRIG_LOW:
+				gpio_cfg.pol_type = FALL_LOW;
+				break;
+			case GPIO_INT_TRIG_HIGH:
+				gpio_cfg.pol_type = RISE_HIGH;
+				break;
+			case GPIO_INT_TRIG_BOTH:
+				return -ENOTSUP;
+			}
+		}
+
+		/* Set IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_cfg.ucGpioNum] = BIT((uint8_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= data->gpio_irq_num[gpio_num];
+	}
+
+	return 0;
+}
+
+static void gpio_eos_s3_isr(const struct device *dev)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	uint8_t intr_status = INTR_CTRL->GPIO_INTR;
+
+	for (uint8_t i = 0; i < MAX_GPIOS; i++) {
+		uint8_t intr_num = BIT(i);
+
+		if (intr_status & intr_num) {
+			for (int gpio_num = 0; gpio_num < MAX_GPIOS; gpio_num++) {
+				if (intr_num == data->gpio_irq_num[gpio_num]) {
+					gpio_fire_callbacks(&data->callbacks,
+							    dev, BIT(gpio_num));
+				}
+			}
+			INTR_CTRL->GPIO_INTR |= intr_num;
+		}
+	}
+}
+
+static const struct gpio_driver_api gpio_eos_s3_driver_api = {
+	.pin_configure = gpio_eos_s3_configure,
+	.port_get_raw = gpio_eos_s3_port_get_raw,
+	.port_set_masked_raw = gpio_eos_s3_port_set_masked_raw,
+	.port_set_bits_raw = gpio_eos_s3_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_eos_s3_port_clear_bits_raw,
+	.port_toggle_bits = gpio_eos_s3_port_toggle_bits,
+	.pin_interrupt_configure = gpio_eos_s3_pin_interrupt_configure,
+	.manage_callback = gpio_eos_s3_manage_callback,
+};
+
+static int gpio_eos_s3_init(const struct device *dev);
+
+const struct gpio_eos_s3_config gpio_eos_s3_config = {
+	.pin_cfgs = DT_INST_PROP(0, pin_cfgs),
+};
+
+static struct gpio_eos_s3_data gpio_eos_s3_data;
+
+DEVICE_DT_INST_DEFINE(0,
+		    gpio_eos_s3_init,
+		    device_pm_control_nop,
+		    &gpio_eos_s3_data,
+		    &gpio_eos_s3_config,
+		    POST_KERNEL,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEFAULT,
+		    &gpio_eos_s3_driver_api);
+
+static int gpio_eos_s3_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority),
+		    gpio_eos_s3_isr,
+		    DEVICE_DT_INST_GET(0),
+		    0);
+
+	irq_enable(DT_INST_IRQN(0));
+
+	return 0;
+}
```

Comment:
It's a non written, de-facto standard to place device instance code at the bottom of the driver file. Let's move this function to just below `static const struct gpio_driver_api gpio_eos_s3_driver_api` where there is currently the declaration.